### DISCUSSION TITLE
Gtk4 Rewrite

### DISCUSSION THREAD

**Discussion by danirabbit** [elementary/photos#713] [General]
We're starting to port apps to Gtk4 now and one I'm really not looking forward to is Photos. This is a really old and extremely complex code base and it still heavily uses things that have been deprecated for a long time. Something I've been thinking about is rethinking the scope of Photos and starting fresh with a more focused goal, so I'd like to start a discussion about what people feel the scope/purpose of the default photos app is and isn't to determine if that's realistic or it would be better to try to port the existing code base

### Things I'd like specific feedback on:
* Photos currently offers some advanced features like raw photo editing. Would it be better to leave advanced photo editing features to more specialized tools like GIMP or Darktable?
* What are the most important tools and workflows to support in photos? Where are photos being imported from and what are they used for? (An example of a workflow would be importing photos from a DSLR, choosing the best photos between multiple similar shots, crop, adjust levels, upload to a file sharing service)
* Some features seem gimmicky (like slideshows). If you had to remove a feature, which would it be?
* What essential features are missing? Filters? Markup? If you could add just one feature, what would it be?
* How do you expect to navigate and organize your photo library? By day? location? person? source of the import (eg screenshots vs camera)?
* Should photos attempt to organize non-photo images like screenshots or wallpapers?

**Comment by r4j4-s** [comment]
Right now my use of the default Photos application is limited. I use it to view images and its relevant metadata, and make very simple changes(rotation, cropping). Never used it for raw photo editing (darktable/rawtherapee). Also never used it to view an entire folder full of photos(also darktable), nevertheless when I tried just now to import photos from a different drive the application unfortunately crashed immediately. Filters would be a nice feature, also would be nice to have blur, annotations and free form select for editing screenshots(would also be nice if it could be integrated into the Screenshots work flow, meaning I need not use ksnip). Navigation using date and person would be useful. I almost never have location data on the photos I take using my camera, and those are mostly the ones I would open on my desktop(photographs from my phone which do have location data go directly to google photos).

**Comment by DimaMzk** [comment]
I think the only thing I've ever used photos for has been:
- viewing
- cropping & rotation
- sometimes adjusting brightness and contrast

A couple of features I'd like to see would be some basic markup features like text and drawing as well as some kind of integration with the screenshot and screen snipping tools that elementary has.

If I ever need to do anything other than that, I'll probably end up opening a dedicated tool like Gimp anyways.

**Comment by MobileAZN** [comment]
I like to use Photos for organization and fast edits, such as cropping, rotating, annotating.  Maybe a basic filter here or there (bonus points for plugins/extensible).  I don't use anything like slideshows or such.  Any advanced editing, I'd use darktable or gimp for.  It would be cool if you could (again, extending) right click the image and "edit in..." where it would then be passed back- almost like Picassa or iPhoto did.  I don't know that it needs to figure out that I have screenshots or wallpapers.

But overall, at the core, I just would like a fast, accessory-like Photos app.  Advanced stuff can be handled by the big apps that already exist for this purpose.  Photos should just be a decent, basic, application.

**Comment by notanimposter** [comment]
I don't use Photos because it doesn't work with small images. It has a maximum zoom level and uses interpolation, so it's useless for things like icons and pixel art. If it could work as well for tiny images as it does for big images, it'd have everything I need (as a basic image previewer), so that'd be my "one feature".

**Comment by treppenwitz03** [comment]
I will be sharing my thoughts as a student who, in this pandemic, uses screenshot utility as a new way of "note-taking" (teachers won't share powerpoints ). 

1. For me, rotating and cropping are the only things I prefer to do with a photo managing application. The more advanced things (at least for me) like adjusting the brightness, or contrast are things I normally would just do in a photo editor. 
2.  As for the workflows, since I am a student, I take so many screenshots and it is easy to lose track of them if I did not label them after a subject. So I think the ability to group photos is good.
3. I think a fair feature to have is ocr integrated, just like how google photos has google lens integrated. I use them when I am making reviewers by feeding what I screenshot-ed to it and funneling the extracted text to a text or word file.
4. Another good feature I think is to be able to print them directly from the photos app.
5. When it comes to navigating the photos, I prefer time and date because it allows me to know what subject was going on when this photo was taken.

I personally think that rewriting it is better, as I find myself preferring to manage photos with the file manager more often than the photos app itself. Just my thoughts :)

**Reply by treppenwitz03** [reply]
https://user-images.githubusercontent.com/73321827/178101933-5cc44cb7-042a-4cf6-ba39-fc218d0fe7b3.mp4

In my practicing, I tried to make it as professional as possible so it can serve as basis of some sort if possible😅 But, I would be honored to have it really adopted..

If you'd like to look at the code, it is at https://github.com/ZenitsuDev/album 

credit is also due to Alexander Mikhaylenko's transition and animation work.

**Reply by BujuArena** [reply]
This doesn't look like it has any way to browse through the images, or at least that wasn't demonstrated.  It shows the user going back to the gallery for every single image they want to view, instead of just going directly to the next and previous image.  It also has large gaps between the thumbnails, and has slow (non-instant) animations for doing things, making browsing the images a bit cumbersome, compared to something like IrfanView which is much quicker.

**Reply by treppenwitz03** [reply]
Well, it was just a prototype I made and could definitely do lots of improving. However, it aims to demonstrate the most basic operation of a photo library, which is to be able to view photo from a library.

**Reply by BujuArena** [reply]
Thanks for your effort.  I'm sorry if my criticism came across as rude.  I certainly appreciate your achievement of opening images and displaying them with the correct colors, aspect ratio, etc.  It's more than can be said for any project I've made from scratch.  I look forward to improvements in usability with this solid foundation.

**Reply by treppenwitz03** [reply]
Worry not! Thanks for pointing out points of improvement

**Comment by BujuArena** [comment]
Copy many aspects of IrfanView and you'll have something incredibly awesome: an easy-to-use, lean, fast image viewer which can view all the images in a directory, take an image pasted from the clipboard and save it however you want, draw some simple shapes and lines if necessary, crop, rotate, resize with choice of algorithm, swap color channels, count unique colors, and see detailed image properties and metadata.  A photo viewer should do at least that and not too much more.  Really, just refer to IrfanView, a very-unfortunately proprietary masterpiece.

**Comment by ovib** [comment]
Most of the time I use the Photos just to **view, crop and rotate** pictures/screenshots. My guess is that this is the case for the majority of the people, hence most of the effort should be put in making this basic features perfect! 

**One frustrating thing regarding cropping and rotating** is that I often need to use them together: I open an image, decide to crop it, I notice that it is not perfectly straight so I rotate it a little bit, but now I need to crop it again and so on... the frustrating part is that to do it on the current Photos app is that I have to switch back and forth between the cropping and rotating tool. As an example, take every default photo app on mobile phones, they allow you to do this two tings in a very intuitive way, at the same time. 
On a desktop, the macOS implementations inside the Photos app is fantastic. Cropping and rotating functionalities are merged: 

![Screenshot from 2022-04-27 13-15-42](https://user-images.githubusercontent.com/38936570/165512299-b7cc23f2-7196-4eba-a1ad-1c19a8422a43.png)

but also having a a vary basic rotation tool, like the one available on Akira for example, WHILE cropping would be still great. 

Regarding **raw photo editing** i sometimes do some basic photo editing (adjusting exposure, contrast etc.) and I generally appreciate if the app handles raw files since the final result is better when using those.

Regarding **navigation / organization / workflow** I don't use a specific app at all. I just open the file I'm interested in from the file system. One thing that I would appreciate though would be to **view also the others pictures in the same folder as the currently opened one** (similarly as in the Preview app on macOS or in a bottom bar like in Adobe Lightroom). 

Regarding **annotations**, I'm a student and I need quite often to add text and/or draw shapes on screenshots (as I did on the image above!). To do that I found the [Annotator app](https://appcenter.elementary.io/com.github.phase1geo.annotator/) by Trevor Williams which works great (@ZenitsuDev take I look at it if you don't know it, you may find it very useful :smiley:  ). For this reason I don't feel the need to have the same functionality in the Photos app, but I can see how for some it might be more convenient.

**Reply by treppenwitz03** [reply]
Cool! I'll take a look at it. Thank you

**Comment by HenryRabelo** [comment]
Helloo! Regarding the Photos app, I think it is really nice to be able to just open the app and see the collection of all your photos, apps that aren't useful when directly opening them are a bit wasteful in my opinion.

I can't really comment on how many people use the Raw photo functionality, although when it comes to user experience, I think it is nice to have a centralized hub to view photos, even camera ones. I also don't know how many people actually use the slideshow feature, neither how much work it would be to maintain it, but it is a neat feature.

If you had to remove a feature, I think the Raw photo editing might be a good candidate, it seems a little excessive if the goal of the app is to be a simple picture viewing app with some simple editing involved, and if someone needs this feature, they are more than likely already using Darktable in their workflow. I do think that importing photos from cameras and viewing Raw photos are great features, however.

I believe that it would be a very welcome addition to add the possibility of navigating the home file system, to filter the photos shown in the app. Many people have folders inside the Images home directory, and especially if there are many pictures to be loaded, the app can get a bit chaotic / overwhelming to navigate and find specific files, even with the automatic dates. Sometimes you just don't need to see all of your pictures at the same time. Another thing it could be nice to add is image metadata viewing / cleaning.

One thing that is a bit confusing is the slight differences in UI and functionality when opening a photo through the app directly vs opening a photo through the Files app. If you are used to cropping images, then right clicking and exporting it to save it, you might be thrown off by the right click export option not being available when opening a picture through the Files app. Likewise, if you are used to just using the "Save Picture With a Different Name" option to export your changes, you might not even be aware that an "Export" option exists when using the app directly. That also applies to deleting a photo when opening it directly: No "Delete" option, you have to press the delete key.

Overall, I do think it is a fantastic app, although it could use a little bit more of consistency and navigability. Sorry if the text ended up a bit long winded, I'm not sure if I should treat the discussion as a thread or not. :)

**Comment by Romchec** [comment]
The only thing i miss, is the ability to fullscreen photo, without slideshow.
P. S. Just found, that you can do it with F11 key, but it will be nice to see button to do it with a signature that you can do it with a key.
Also i wanted to see trackpad integration with pinch-to-zoom and ability to drag photo with 2 fingers.

**Comment by zeebok** [comment]
I do not use Photos a huge amount for raw editing at the moment but I am probably a bit of an outlier in that my typical workflow is only a couple features beyond what Photos currently does and so would be pretty happy to get the last couple of them, and just use Photos solely since it generally feels much snappier and cleaner to me.

For organizing I would need it to just watch a folder, let me do some rating by number or color, and let me group them based on a named event or just stay with the default dates. For editing I already said that what it has is close to what I would want for raw editing, just would like local contrast and some color channel control.  I think those would be the really essential features to add to have me stick with using the app. With exporting, I don't think the uploading to a service is really necessary or maybe to me the selection is just outdated feeling? If I wanted to upload my photos somewhere it would be something like SmugMug or Squarespace or something. I wonder if this is something that Online Accounts is meant for?

I think slideshows and support for videos is not really needed and can be tossed. If keeping the service uploading up to date is hard, I don't think it would be a huge loss personally. I find the raw file thumbnail generation very frustrating as I have Nextcloud watch the folder and don't want it to get riddled with thumbnails when I believe most modern raw formats provide a preview image in them. Or perhaps other applications just store them in a hidden folder elsewhere.

That said I don't think it would be unreasonable to scale back the scope of what the app's core functionality is and wouldn't be disappointed if it meant a cleaner code base that doesn't make eOS devs shudder when they take a look at it!

**Reply by zeebok** [reply]
I think it is important to consider what Photos should accomplish in other form factors. If eOS ends up on a tablet or mobile device, things like darktable become much less viable as a solution for raw editing. Should Photos provide something basic like you find on iOS and Android?

**Comment by Maccer1** [comment]
Maybe an idea... Can we have an estimation on how difficult a feature is to implement and then we can decide if it's really worth it?

I use a lot of things about the photos app and so everyone else. So at least this way we can have a better understanding on what the difficult points are.

**Reply by Maccer1** [reply]
Some examples of functionality:
- Modify metadata
- Crop images
- Adjust brightness/contrast/gamma...
- Save pictures to a cloud drive
- Organize pictures taking info from metadata (year, location, camera...)
- Make backups of pictures
- Create a library of pictures under a directory
- Import pictures from MTP/PTP devices (cameras/phones)
- Import pictures from USB drives/SD cards
- etc

**Comment by bzx0** [comment]
> * What essential features are missing?

1. Show size of image
2. Navigate back/forward with mouse scroll/touchpad gesture

**Comment by TomiOhl** [comment]
> Photos currently offers some advanced features like raw photo editing. Would it be better to leave advanced photo editing features to more specialized tools like GIMP or Darktable?

Absolutely. Tbh I didn't know about this feature in Photos, but probably I'd use a photo editing app for such a complex task.

> What are the most important tools and workflows to support in photos? Where are photos being imported from and what are they used for? (An example of a workflow would be importing photos from a DSLR, choosing the best photos between multiple similar shots, crop, adjust levels, upload to a file sharing service)

How I use Photos:
- opening photos from a folder (in file explorer) to preview them. I expect to be able to navigate back and forth between the photos in the same folder.
- open the app itself to take a look at my library. The library itself should just be a (set of) selected folder(s) and it's/their subfolder(s). 
- cropping and rotating photos are the only editing features I use. For more advanced editing, I use specific software.
- on the preview screen, apart from the mentioned editing features I look for the metadata, file data, delete and set as wallpaper options.

> Some features seem gimmicky (like slideshows). If you had to remove a feature, which would it be?

I definitely never use the mentioned slideshows feature or the Facebook/Flickr/etc integrations.

> What essential features are missing? Filters? Markup? If you could add just one feature, what would it be?

Pinch to zoom and swipe to navigate between photos (like with arrow keys)! These would make the experience with a touchpad/touchscreen so much better!

> How do you expect to navigate and organize your photo library? By day? location? person? source of the import (eg screenshots vs camera)?

I prefer chronological ordering + folders view.

> Should photos attempt to organize non-photo images like screenshots or wallpapers?

I wouldn't differentiate different kinds of images from others.

**Comment by hanaral** [comment]
Just create something that fits the function of viewing one's photo library and albums in an intuitive and simple way like MacOS photos. Being able to import photos from DCIM is a must, only really because it requires the Photos app itself to index items.

**Comment by peteruithoven** [comment]
Features:
- Viewing (Basically only from Files, never the library/gallery mode)
- Cropping
- Rotation
- Deleting (See: #44)

I only use it to open photos / screenshots from Files. So when going to the prev/next image I would really appreciate it if it could follow the current image's folder's sorting, see: #14.

**Comment by brownphotographic** [comment]
Very exciting! A few ideas:

### Tenets:
- Simple to use, responsive... as per the original objectives behind the Yorba project that was Shotwell
- Focus on the basics, not bloated features like digikam has

### MVP Features (in no particular order):

1. **Ratings / reject for image curation** - star rating (0-5 plus reject) is the standard. This is the only reason I still use Shotwell and not Elementary Photos.
2. **Events / Albums**... let's call these Albums. Auto generate by date taken, allow user to edit as per current. In addition, allow user to create new albums and 'folders' of albums. This allows for the curation of a collection of images that is separate to how they are stored on the disk.
3. **Auto import from > 1 'watch folder'**. Currently limited to just one. Let user add many, and choose which one to import to. This would allow support for screenshots, storing images on different disks etc.
4. **Raw and JPEG support** (and any other image format). For RAW show the embedded JPEG. Also continue to treat RAW and JPEG as the same image... ie stack them together so one doesn't see RAW and JPEG as separate options.... BUT when I say RAW support don't go crazy at first and make this a best in class raw editor. There are other apps (RAW Therapee, Digikam) to do that in... and this is a huge rat hole! Just show the embedded JPEG and be done with it.
6. **Open with** (of the user's choice... e.g. GIMP, Rawtherapee): External Editor and RAW editor (as two independent options). 
7. Very **basic image editing** per current SW (just exposure, contrast, saturation etc). Let the heavy lifting be driven by other RAW editors. Per the current functionality, allow RAW images to be 'edited' but just make the adjustments based on the embedded JPEG.
8. **Crop, straighten, rotate**
9. The above two with **non destructive editing**. Ideally allow the user to create a new version of an image when they make the edit, as opposed to overwriting the original. Also allow metadata to be non-destructive (ie allow it to live either in the library, or an XMP)
10. **Print** function (just use the system dialog)
11. Continue to support **export plugins** via Publish button... flickr, google photos, dropbox etc (perhaps enable this to be a community driven effort?). Perhaps there are already open source options for these publish options so you don't have to write everything.
12. **Search**
13. **Hierarchical tagging**
14. All functions accessible via **keyboard shortcuts**
15. **Color management** - just at least respect the color profile assigned at the OS level.
16. View image metadata (camera settings, other image metadata)
17. Of course a **published flatpak** so the whole linux community can use
18. Continue to support **videos** as they are

### Nice to have features:
- **'Stacks' of images for organizing** RAW + JPEG, new image versions etc. Just do it exactly like this! https://www.youtube.com/watch?v=t18dRKU7kb8
- Allow user to **relocate the images on disk**... so if one needs to move an image to different disks, it can be done with the app being aware of the move. https://www.youtube.com/watch?v=fDvzUPw8ME0
- **XMP sidecar** to store metadata of an image (tags, edits)... this would allow for inter-app interoperability.


### Ditch:
- Slideshow (but keep fullscreen)
- Red eye removal

**Comment by Clind99** [comment]
First thanks for your work and I'm very pleased by the possibility to have my favourite Photo manager refreshed and enhanced :D
Here are a list of features I mainly use and other I've never used in Photos... 
Hop this help !

1 Incremental import and watch auto-import (as it is)
2 Rating (stars and reject automatically masking) 
3 Flagging (use it to select files for the next working session)
4 RAW/Jpeg grouping under a same entity (XMP management for export and deletion would be a plus)
5 Event detection as it is already present
6 Hierarchycal TAGS
7 Timeline with amount of shot taken and possibility to restrain mosaic view/selection to a time frame with it
8 quick search field as it is where filenames/extensions tags can be mixed (maybe adding other metadatas ? body name ? Focal lenth ?)
9 Export a selecion with or without changing format, resolution or including metadata (as it is)
10 Possibility to write tags to metadata
11 possibility to edit all metadata (single or group)
12 Menu to open in external raster or RAW editor
13 Video support as it is is fine
14 Quick filetype and rating filters as they are are convinient

Features I do not use personally
1 Sideshows
2 Export plug-ins
3 Edition/Correction features (level, crop, enhance, red-eyes) as I use RAW editor (Darktable) or Raster (Gimp) to work on my images
 

1 Incremental import and watch auto-import (as it is)
2 Rating (stars and reject automatically masking) 
3 Flagging (use it to select files for the next working session)
4 RAW/Jpeg grouping under a same entity (XMP management for export and deletion would be a plus)
5 Event detection as it is already present
6 Hierarchycal TAGS
7 Timeline with amount of shot taken and possibility to restrain mosaic view/selection to a time frame with it
8 quick search field as it is where filenames/extensions tags can be mixed (maybe adding other metadatas ? body name ? Focal lenth ?)
9 Export a selecion with or without changing format, resolution or including metadata (as it is)
10 Possibility to write tags to metadata
11 possibility to edit all metadata (single or group)
12 Menu to open in external raster or RAW editor
13 Video support as it is is fine
14 Quick filetype and rating filters as they are are convinient

Features I do not use personally
1 Sideshows
2 Export plug-ins
3 Edition/Correction features (level, crop, enhance, red-eyes) as I use RAW editor (Darktable) or Raster (Gimp) to work on my images

**Comment by DSLitvinov** [comment]
I was pleasantly surprised by the functionality of Photos, very similar to a reference [app ](https://eagle.cool/).
What I don’t use: 
1. Export plug-ins
2. 2 Sideshows

What is missing:
Capture the image color palette and search images by color.
A large field for tags, now it is very tight and inconvenient.
